### ISSUE TITLE
Performance comparison with standard PriorityQueue

### ISSUE THREAD

**Issue by karussell** [michaeltandy/contraction-hierarchies#1]
Did you had a chance to compare the speed when using `java.util.PriorityQueue` instead of the FibonacciHeap?

**Comment by karussell** [comment]
Using FibonacciHeap makes my bidirectional algorithm roughly 10% slower when using CH, without CH it is even over 40% slower with FibonacciHeap. I would be curious which numbers you get.

**Comment by michaeltandy** [comment]
Before I adopted the FibonacciHeap I used the standard Java Collections TreeMap. Fibonacci heap has O(log n) remove-minimum performance but O(1) decrease key performance, whereas the TreeMap has O(log n) performance for both operations. On the other hand, being core java TreeMap might have got more JIT testing. Like TreeMap, PriorityQueue has O(log n) performance on all the operations we're interested in.

The benchmark I did before deciding to keep the FibonacciHeap reduced from 74 seconds to 63 seconds with the introduction of the FibonacciHeap. I've since decided that wasn't really the best benchmark. My current benchmark can be found in BenchmarkUk.java - I test using OpenStreetMap data for the UK, and path between Hatfield and 4,000 randomly chosen nodes.

With FibonacciHeap:
2 repetitions uncached pathing from hatfield to 4000 locations in 152582 ms.
2 repetitions parallel uncached pathing from hatfield to 4000 locations in 102126 ms.

I ran the same benchmark using TreeMap and two implementations with very minor differences:
2 repetitions uncached pathing from hatfield to 4000 locations in 143536 ms.
2 repetitions parallel uncached pathing from hatfield to 4000 locations in 98000 ms.

2 repetitions uncached pathing from hatfield to 4000 locations in 138642 ms.
2 repetitions parallel uncached pathing from hatfield to 4000 locations in 100675 ms.

I put together an implementation using PriorityQueue to see what would happen:
2 repetitions uncached pathing from hatfield to 4000 locations in 130723 ms.
2 repetitions parallel uncached pathing from hatfield to 4000 locations in 91206 ms.

In summary I guess the Fibonacci heap doesn't seem to be an improvement, and it might actually be harming performance. And despite TreeMap and PriorityQueue having the same asymptotic performance, PriorityQueue seems to perform marginally better in practice.

**Comment by michaeltandy** [comment]
I did some more performance tests, this time using some r3.xlarge EC2 instances. With 22704cecbf9d0a8109f76278580ec4ef9c475e84 the average time was 152673ms and the max time was 179975ms - that's a pretty good improvement over aa0773713aec296f460dda1610cf7a1cb8689b09 where the average was 185775ms and the max time was 215830ms

**Comment by karussell** [comment]
Really interesting that you get the same boost! Thanks!
